nkms_eth/escrow.py

```diff
@@ -1,78 +1,132 @@
+            raise self.ContractDeploymentError(message)
+
+        contract, deploy_txhash = self.blockchain.chain.provider.deploy_contract(self._contract_name,
+                                                          deploy_args=[self.token.contract.address] + self.mining_coeff,
+                                                          deploy_transaction={'from': self.token.creator})
+
+        self.blockchain.chain.wait.for_receipt(deploy_txhash, timeout=self.blockchain.timeout)
+        self.contract = contract
+
+        reward_txhash = self.token.transact({'from': self.token.creator}).transfer(self.contract.address, self.reward)
+        self.blockchain.chain.wait.for_receipt(reward_txhash, timeout=self.blockchain.timeout)
+
+        init_txhash = self.contract.transact({'from': self.token.creator}).initialize()
+        self.blockchain.chain.wait.for_receipt(init_txhash, timeout=self.blockchain.timeout)
+
+        return deploy_txhash, reward_txhash, init_txhash
+
+    @classmethod
+    def get(cls, blockchain, token):
+        """Returns an escrow object or an error"""
+        contract = blockchain.chain.provider.get_contract(cls._contract_name)
+        return cls(blockchain=blockchain, token=token, contract=contract)
+
+    def transact(self, *args, **kwargs):
+        if not self.contract:
+            raise self.ContractDeploymentError('Contract must be deployed before executing transactions.')
+        return self.contract.transact(*args, **kwargs)
+
+    def get_dht(self) -> set:
+        """Fetch all DHT keys and return them as a set"""
+        return {miner.get_dht_key()for miner in self.miners}
+
+    def get_swarm(self):
+        miner, i = self.null_addr, 0
+        while True:
+            next_miner = self.__call__().getNextMiner(miner)
+            if next_miner == self.null_addr:
+                raise StopIteration()
+
+            yield next_miner
+            miner = next_miner
+            i += 1
+
+    def sample(self, quantity: int=10, additional_ursulas: float=1.7, attempts: int=5, duration: int=10) -> List[addr]:
+        """
+        Select n random staking Ursulas, according to their stake distribution.
+        The returned addresses are shuffled, so one can request more than needed and
+        throw away those which do not respond.
+        """
+
+        system_random = random.SystemRandom()
+        n_select = round(quantity*additional_ursulas)            # Select more Ursulas
+        n_tokens = self.__call__().getAllLockedTokens()
+
+        if not n_tokens:
+            raise self.NotEnoughUrsulas('No locked tokens.')
+
+        for _ in range(attempts):
+            points = [0] + sorted(system_random.randrange(n_tokens) for _ in range(n_select))
+            deltas = [i-j for i, j in zip(points[1:], points[:-1])]
+
+            addrs, addr, shift = set(), self.null_addr, 0
+            for delta in deltas:
+                addr, shift = self.__call__().findCumSum(addr, delta+shift, duration)
```

Comment:
Very cool.  Maybe a link to the diagram from the whitepaper as a comment (or in the docstring above).  It's cool to see it as python.

Reply:
I have an ascii art in the `.sol` file.. But I think a link to the actual diagram (in both places) is better:
https://github.com/nucypher/kms-whitepaper/blob/master/pdf/miners-ruler.pdf